Allow tasks to succeed with non-completed outputs.

hjoliver opened the pull request:
We should be able to trigger alternate workflows with alternate message triggers (in fact I need this for a suite I'm working on now). This is not currently possible because all task outputs are forced to the completed state when a task finishes.

This change stops the auto-completion of message outputs if a task succeeds. 
 
Side-effects: message outputs will no longer be auto-completed:
 1. when a task succeeds, if the earlier task message failed to get through (e.g. due to a temporary network outage)
 2. at restart, when an orphaned task with message outputs polls as succeeded 
 3. when a task state is manually set to 'succeeded'.

The right way to handle 1. is to poll the job status file for outputs (a) if a task succeeds with incomplete outputs, or (b) for orphaned tasks at restart: #1549.  

The right way to handle 3. is allow "cylc reset" to set the state of individual message outputs if necessary, because generally it's not valid to assume anything about them: #1550.

hjoliver commented:
On second thoughts, to avoid affecting others before #1549 and #1550 are done I've restored current behaviour for polling as succeeded (e.g. at restart) and manual reset to succeeded.  So current suites will only be affected in the unlikely event that a message output fails to get through but the succeeded message does (or if the message output was never being reported in the first place, but that's a bug).

hjoliver commented:
@matthewrmshin - please review.

hjoliver commented:
Note that the 'submitted' and 'started' standard (non-message) outputs are still auto-completed.  This was originally intended for fast-running tasks that finish before job submission is reported, which used to be possible (I'm not sure if it still is).  I think it would be simpler not to do this and just complete the 'submitted' output when submission is reported out of order ... but that's now another issue: #1552.

matthewrmshin commented:
I cannot get these tests to pass at the moment. Is it because they all use custom outputs?

```
./tests/cyclers/40-integer1.t
./tests/database/04-lock-recover.t
./tests/database/05-lock-recover-100.t
```

hjoliver commented:
Ah, yes .. didn't get around to running the test battery!  later...

hjoliver commented:
Tests fixed.

matthewrmshin commented on tests/database/04-lock-recover/suite.rc:
```diff
@@ -12,6 +12,7 @@ locker:lock => TRUES:succeed-all => done
     [[locker]]
         script = """
 timeout 60 cylc-db-lock
+cylc message "$CYLC_TASK_CYCLE_POINT I have locked the public database file"
```
`cylc-db-lock` calls `cylc task message "[1] I have locked the public database file"` in Python. Have I done this wrong when I wrote this test? Ditto for the test 05.

hjoliver replied:
Yeah, it's wrong; the square brackets are placeholder for cycle point with no offset, not literal.  It only worked because of cylc automatically completing non-completed outputs when the task finished (and this branch doesn't do that).

matthewrmshin replied:
In which case, I think we need to fix the message call in `cylc-db-lock`?

matthewrmshin commented:
@arjclark please review 2.

hjoliver commented:
> In which case, I think we need to fix the message call in cylc-db-lock?

I've fixed the original message output strings in the Python db-lock scripts, and removed my new in-suite replacements.

matthewrmshin commented:
All looking good to me.

hjoliver commented:
Branch updated from master. Test battery still passes.

arjclark commented:
Can't seem to get this one to pass:

./tests/database/05-lock-recover-100.t (Wstat: 0 Tests: 3 Failed: 2)
  Failed tests:  2-3

hjoliver commented:
@arjclark - the ```timeout``` in the ```locker``` command scripting wasn't long enough.  (something appears to have slowed things down a bit, maybe the new authentication...)

hjoliver commented on tests/database/05-lock-recover-100/bin/cylc-db-lock:
```diff
@@ -11,12 +12,14 @@ def main():
         for line in open(os.path.join(suite_log_dir, "log")):
             if "cylc-suite.db: recovered from" in line:
                 return
+        time.sleep(1)
```
This loop was hammering the CPU.

arjclark commented:
Test now passing.
